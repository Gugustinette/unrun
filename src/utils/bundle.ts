import { randomBytes } from 'node:crypto'
import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import { pathToFileURL } from 'node:url'
import {
  build,
  type InputOptions,
  type OutputOptions,
  type RolldownOutput,
} from 'rolldown'
import {
  createMakeCjsWrapperAsyncFriendlyPlugin,
  createRequireResolveFix,
  createRequireTypeofFix,
  createSourceContextShimsPlugin,
} from '../plugins'
import type { ResolvedOptions } from '../options'

export async function bundle(options: ResolvedOptions): Promise<{
  outDir: string
  rolldownOutput: RolldownOutput
}> {
  // Resolve the file path to an absolute path
  options.path = path.resolve(process.cwd(), options.path)
  // Ensure output directory exists
  const unrunOutDir = path.join(process.cwd(), 'node_modules', '.unrun')
  fs.mkdirSync(unrunOutDir, { recursive: true })
  // Compute key for output filename
  const moduleKey = randomBytes(16).toString('hex')
  // Create directory dedicated to this module to avoid collisions
  const outDir = path.join(unrunOutDir, moduleKey)
  fs.mkdirSync(outDir, { recursive: true })

  // Input options (https://rolldown.rs/reference/config-options#inputoptions)
  const inputOptions: InputOptions = {
    input: options.path,
    // Use Node platform for better Node-compatible resolution & builtins
    // See https://rolldown.rs/guide/in-depth/bundling-cjs#require-external-modules
    platform: 'node' as const,
    // Treat all non-relative and non-absolute imports as external dependencies,
    //   except for package "imports" specifiers (starting with '#') which we resolve ourselves.
    //   This ensures bare deps resolve from host project while allowing #imports mapping.
    external: (id: string) =>
      !id.startsWith('.') && !id.startsWith('/') && !id.startsWith('#'),
    // Keep __dirname/__filename/import.meta.url definitions
    define: {
      __dirname: JSON.stringify(path.dirname(options.path)),
      __filename: JSON.stringify(options.path),
      'import.meta.url': JSON.stringify(pathToFileURL(options.path).href),
      'import.meta.filename': JSON.stringify(options.path),
      'import.meta.dirname': JSON.stringify(path.dirname(options.path)),
      'import.meta.env': 'process.env',
    },
    // Compose feature-specific plugins
    plugins: [
      // Inject __dirname/__filename/import.meta shims and inline import.meta.resolve
      createSourceContextShimsPlugin(),
      // Make CJS wrappers async-friendly
      options.makeCjsWrapperAsyncFriendly
        ? createMakeCjsWrapperAsyncFriendlyPlugin()
        : null,
      // Fix require.resolve calls to use correct base path
      createRequireResolveFix(options),
      // Fix typeof require in ESM
      createRequireTypeofFix(),
    ],
    // Resolve tsconfig.json from cwd if present
    tsconfig: path.resolve(process.cwd(), 'tsconfig.json'),
    keepNames: true,
    // Configure JSX support with classic mode for pragma support
    jsx: {
      mode: 'classic',
      factory: 'React.createElement',
      fragment: 'React.Fragment',
    },
    // Finally, apply user-provided overrides
    ...options.inputOptions,
  }

  // Output options (https://rolldown.rs/reference/config-options#outputoptions)
  const outputOptions: OutputOptions = {
    dir: outDir,
    format: 'esm',
    // Apply user-provided overrides
    ...options.outputOptions,
  }

  // Let Rolldown handle writing the bundle to disk
  const rolldownOutput = await build({
    ...inputOptions,
    output: outputOptions,
  })

  // Return the output file path
  return {
    outDir,
    rolldownOutput,
  }
}
