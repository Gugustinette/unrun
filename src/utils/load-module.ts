import { Buffer } from 'node:buffer'
import crypto from 'node:crypto'
import fs from 'node:fs'
import { tmpdir } from 'node:os'
import path from 'node:path'
import process from 'node:process'
import { pathToFileURL } from 'node:url'

/**
 * Import a JS module from code string.
 * Write ESM code to a temp file (prefer project-local node_modules/.unrun) and import it.
 * Cleans up file unless keepFile is true.
 * @param code - The JavaScript code to be imported as a module.
 * @param opts - Optional parameters.
 * @param opts.filenameHint - A hint for the filename to use when creating the temp file.
 * @param opts.keepFile - If true, the temp file will not be deleted after import.
 * @returns The imported module.
 */
export async function loadModule(
  code: string,
  opts: { filenameHint?: string; keepFile?: boolean } = {},
): Promise<any> {
  let moduleUrl: string | null = null
  const { filenameHint, keepFile } = opts
  try {
    const hash = crypto.createHash('sha1').update(code).digest('hex')
    const fname = `${filenameHint ? `${sanitize(filenameHint)}.` : ''}${hash}.mjs`
    const projectNodeModules = path.join(process.cwd(), 'node_modules')
    const outDir = path.join(projectNodeModules, '.unrun')
    const outFile = path.join(outDir, fname)
    if (!fs.existsSync(outFile)) {
      try {
        fs.mkdirSync(outDir, { recursive: true })
        fs.writeFileSync(outFile, code, 'utf8')
      } catch {
        const fallbackDir = path.join(tmpdir(), 'unrun-cache')
        const fallbackFile = path.join(fallbackDir, fname)
        fs.mkdirSync(fallbackDir, { recursive: true })
        fs.writeFileSync(fallbackFile, code, 'utf8')
        moduleUrl = pathToFileURL(fallbackFile).href
      }
    }
    moduleUrl = moduleUrl || pathToFileURL(outFile).href
  } catch {
    moduleUrl = `data:text/javascript;base64,${Buffer.from(code).toString('base64')}`
  }

  let _module
  try {
    _module = await import(moduleUrl)
  } finally {
    if (!keepFile && moduleUrl.startsWith('file://')) {
      try {
        fs.unlinkSync(new URL(moduleUrl))
      } catch {}
    }
  }
  return _module
}

function sanitize(name: string) {
  // Allow word chars plus dot and dash
  return name.replaceAll(/[^\w.-]/g, '_')
}
