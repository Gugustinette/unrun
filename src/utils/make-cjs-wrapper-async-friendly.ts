/**
 * Transforms code strings containing CommonJS wrappers to be async-friendly.
 *
 * Rolldown may wrap CommonJS modules in a `__commonJS` function that uses
 * arrow functions. If the wrapped code contains top-level `await`, this can lead
 * to syntax errors since the callback function won't be marked as `async`.
 * This function scans for such patterns and modifies the arrow functions to
 * be `async` if they contain `await` expressions.
 * @param input - The input code string potentially containing CommonJS wrappers.
 * @returns The transformed code string with async-friendly wrappers.
 */
export function makeCjsWrapperAsyncFriendly(input: string): string {
  let code = input
  const marker = '__commonJS({'
  if (!code.includes(marker)) return input

  let pos = 0
  let changed = false
  const arrowToken = '(() => {'
  const asyncArrowToken = '(async () => {'

  while (true) {
    const markerIdx = code.indexOf(marker, pos)
    if (markerIdx === -1) break

    const fnStart = code.indexOf(arrowToken, markerIdx)
    if (fnStart === -1) {
      pos = markerIdx + marker.length
      continue
    }

    const bodyStart = fnStart + arrowToken.length
    let i = bodyStart
    let depth = 1
    while (i < code.length && depth > 0) {
      const ch = code[i++]
      if (ch === '{') depth++
      else if (ch === '}') depth--
    }
    if (depth !== 0) break

    const bodyEnd = i - 1
    const body = code.slice(bodyStart, bodyEnd)

    if (
      /\bawait\b/.test(body) && // Only insert if not already async
      code.slice(fnStart, fnStart + asyncArrowToken.length) !== asyncArrowToken
    ) {
      code = `${code.slice(0, fnStart + 1)}async ${code.slice(fnStart + 1)}`
      changed = true
      pos = fnStart + 1 + 'async '.length
      continue
    }

    pos = bodyEnd
  }

  return changed ? code : input
}
