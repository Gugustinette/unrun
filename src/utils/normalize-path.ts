import { fileURLToPath } from 'node:url'

/**
 * Normalize a path-like input to a string path.
 * @param pathLike The path-like input (string or URL).
 * @returns The normalized string path.
 */
export function normalizePath(pathLike?: string | URL): string {
  if (!pathLike) {
    return 'index.ts'
  }

  if (pathLike instanceof URL) {
    if (pathLike.protocol === 'file:') {
      return fileURLToPath(pathLike)
    }
    return pathLike.href
  }

  if (typeof pathLike === 'string') {
    if (!pathLike.startsWith('file:')) {
      return pathLike
    }

    try {
      return fileURLToPath(pathLike)
    } catch {
      try {
        return fileURLToPath(new URL(pathLike))
      } catch {
        return pathLike
      }
    }
  }

  return String(pathLike)
}
