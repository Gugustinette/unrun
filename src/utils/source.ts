/**
 * Common source helpers: shebang stripping and ESM detection.
 */

/**
 * Remove a hashbang (#!/usr/bin/env node) from the beginning of the source,
 * preserving line numbers as much as possible.
 */
export function stripShebang(src: string): string {
  if (src.startsWith('#!')) {
    const nl = src.indexOf('\n')
    return nl === -1 ? '' : src.slice(nl + 1)
  }
  return src
}

/**
 * A very light heuristic to detect if a module looks like ESM.
 * We avoid false positives by requiring bare `import`/`export` tokens.
 */
export function isEsmLike(src: string): boolean {
  return /\b(?:import|export)\b/.test(src)
}

/**
 * Test if a string is a valid JS identifier (for named JSON exports, etc.).
 */
export function isValidIdentifier(name: string): boolean {
  return /^[a-z_$][\w$]*$/i.test(name)
}
