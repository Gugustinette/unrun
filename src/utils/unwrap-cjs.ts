/**
 * Unwrap rolldown's __commonJS wrappers that contain top-level await.
 *
 * Why:
 * - Rolldown may transform certain modules using __commonJS wrappers. If the
 *   wrapped body contains `await`, keeping the wrapper would make the code
 *   invalid under ESM parsing rules (await not allowed in non-async function).
 *
 * How:
 * - Scan output for patterns: `var <name> = __commonJS({ ... (() => { <body> }) });`
 * - If <body> contains `await`, strip the var assignment and splice in <body>.
 * - Remove any trailing `export default <name>()` calls, and as a heuristic,
 *   also strip `export default require_*()` tails.
 */
export function unwrapCjsWrapper(input: string): string {
  let code = input
  const marker = '__commonJS({'
  // Fast-path: if marker not found, skip any extra work
  if (!code.includes(marker)) return input
  let pos = 0
  let changed = false
  while (true) {
    const varIdx = code.indexOf(marker, pos)
    if (varIdx === -1) break
    const varDeclStart = code.lastIndexOf('var ', varIdx)
    if (varDeclStart === -1) {
      pos = varIdx + marker.length
      continue
    }
    const before = code.slice(varDeclStart, varIdx)
    const m = /var\s+([A-Za-z_$][\w$]*)\s*=\s*$/.exec(before)
    const varName = m?.[1] || null
    const fnStart = code.indexOf('(() => {', varIdx)
    if (fnStart === -1) {
      pos = varIdx + marker.length
      continue
    }
    const bodyStart = fnStart + '(() => {'.length
    let i = bodyStart
    let depth = 1
    while (i < code.length && depth > 0) {
      const ch = code[i++]
      if (ch === '{') depth++
      else if (ch === '}') depth--
    }
    if (depth !== 0) break
    const bodyEnd = i - 1
    const body = code.slice(bodyStart, bodyEnd)
    if (!/\bawait\b/.test(body)) {
      // Skip past this wrapper and continue searching
      pos = bodyEnd
      continue
    }
    // Find the end of the __commonJS wrapper statement (typically '});')
    let wrapperEnd = code.indexOf('});', bodyEnd)
    if (wrapperEnd === -1) {
      const semi = code.indexOf(';', bodyEnd)
      wrapperEnd = semi === -1 ? bodyEnd : semi + 1
    } else {
      wrapperEnd += 3
    }
    // Replace the entire var assignment + wrapper with the unwrapped body
    code =
      code.slice(0, varDeclStart) +
      code.slice(bodyStart, bodyEnd) +
      code.slice(wrapperEnd)
    if (varName) {
      const exportCallRE = new RegExp(
        `export\\s+default\\s+${varName}\\s*\\(\\)\\s*;?`,
        'g',
      )
      code = code.replace(exportCallRE, '')
    }
    // Heuristic: remove any default export calling a CJS wrapper like require_foo()
    code = code.replaceAll(/export\s+default\s+require_[\w$]+\s*\(\)\s*;?/g, '')
    changed = true
    // Continue scanning from just after the inserted body
    pos = varDeclStart + (bodyEnd - bodyStart)
  }
  return changed ? code : input
}
