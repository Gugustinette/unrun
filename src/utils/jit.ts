import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import type { ResolvedOptions } from '../options'
import { bundle } from './bundle'
import { loadModule } from './load-module'
import { makeCjsWrapperAsyncFriendly } from './make-cjs-wrapper-async-friendly'

export const jit = async (options: ResolvedOptions): Promise<any> => {
  // Resolve the file path to an absolute path
  const filePath = path.resolve(process.cwd(), options.path)

  // Check if the file exists
  if (!fs.existsSync(filePath)) {
    throw new Error(`[unrun] File not found: ${filePath}`)
  }

  // Bundle the code
  const outputChunk = await bundle(options)

  // Post-process: make CommonJS wrappers async-friendly
  let finalCode = outputChunk.code
  if (options.makeCjsWrapperAsyncFriendly ?? true) {
    finalCode = makeCjsWrapperAsyncFriendly(finalCode)
  }

  // Load the generated module
  let _module
  try {
    _module = await loadModule(finalCode, {
      filenameHint: path.basename(options.path),
      keepFile: process.env.UNRUN_DEBUG === 'true',
    })
  } catch (error) {
    throw new Error(
      `[unrun] Import failed (code length: ${finalCode.length}): ${(error as Error).message}`,
    )
  }

  // Return the module
  return _module
}
