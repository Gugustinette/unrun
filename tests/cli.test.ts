import { resolve } from 'node:path'
import { describe, expect, test } from 'vitest'
import { runNodeCli, unrunCliEntry } from './utils/run-cli'

const fixturePath = resolve(__dirname, 'fixtures/cli.fixture.ts')

describe('cli', () => {
  test('should display nothing when no command line arguments are given', async () => {
    const { stdout, stderr } = await runNodeCli(unrunCliEntry, [fixturePath])

    // Snapshot the outputs
    expect(stdout).toMatchSnapshot('cli-no-args')
    expect(stderr).toMatchSnapshot('cli-no-args-stderr')
  })

  test('should not display command line arguments before the file path', async () => {
    const { stdout, stderr } = await runNodeCli(unrunCliEntry, [
      '--option-that-should-not-appear',
      fixturePath,
      '--some-arg',
      'value',
      '--another-arg=42',
    ])

    // Snapshot the outputs
    expect(stdout).toMatchSnapshot('cli-with-args')
    expect(stderr).toMatchSnapshot('cli-with-args-stderr')
  })
})
