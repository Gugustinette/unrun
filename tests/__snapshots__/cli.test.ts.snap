// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cli > should display nothing when no command line arguments are given > cli-no-args 1`] = `
"[]
"
`;

exports[`cli > should display nothing when no command line arguments are given > cli-no-args-stderr 1`] = `""`;

exports[`cli > should not display command line arguments before the file path > cli-with-args 1`] = `
"[ '--some-arg', 'value', '--another-arg=42' ]
"
`;

exports[`cli > should not display command line arguments before the file path > cli-with-args-stderr 1`] = `""`;
